packages/connection/src/class-nonce-handler.php

```diff
@@ -0,0 +1,197 @@
+		}
+
+		// This should always have gone through Jetpack_Signature::sign_request() first to check $timestamp an $nonce.
+		$timestamp = (int) $timestamp;
+		$nonce     = esc_sql( $nonce );
+
+		// Raw query so we can avoid races: add_option will also update.
+		$show_errors = $wpdb->show_errors( false );
+
+		$old_nonce = $wpdb->get_row(
+			$wpdb->prepare( "SELECT * FROM `$wpdb->options` WHERE option_name = %s", "jetpack_nonce_{$timestamp}_{$nonce}" )
+		);
+
+		if ( is_null( $old_nonce ) ) {
+			$return = (bool) $wpdb->query(
+				$wpdb->prepare(
+					"INSERT INTO `$wpdb->options` (`option_name`, `option_value`, `autoload`) VALUES (%s, %s, %s)",
+					"jetpack_nonce_{$timestamp}_{$nonce}",
+					time(),
+					'no'
+				)
+			);
+
+			/**
+			 * Use the filter to disable the nonce cleanup that happens at shutdown after adding a new nonce.
+			 *
+			 * @since 9.0.0
+			 *
+			 * @param int $limit How many old nonces to remove at shutdown.
+			 */
+			if ( apply_filters( 'jetpack_connection_add_nonce_cleanup', $run_cleanup ) ) {
+				add_action( 'shutdown', array( __CLASS__, 'clean_runtime' ) );
+			}
+		} else {
+			$return = false;
+		}
+
+		$wpdb->show_errors( $show_errors );
+
+		static::$nonces_used_this_request[ "$timestamp:$nonce" ] = $return;
+
+		return $return;
+	}
+
+	/**
+	 * Removing all the nonces.
+	 *
+	 * @param int $cutoff_timestamp All nonces added before this timestamp will be removed.
+	 *
+	 * @return true
+	 */
+	public static function clean_all( $cutoff_timestamp = PHP_INT_MAX ) {
+		for ( $i = 0; $i < 1000; ++$i ) {
+			$result = static::delete( static::CLEAN_ALL_LIMIT_PER_BATCH, $cutoff_timestamp );
+
+			if ( ! $result ) {
+				break;
+			}
+		}
+
+		return true;
+	}
+
+	/**
+	 * Clean up the expired nonces on shutdown.
+	 *
+	 * @return bool True if the cleanup query has been run, false if the table is locked.
+	 */
+	public static function clean_runtime() {
+		if ( static::is_table_locked() ) {
+			return false;
+		}
+
+		/**
+		 * Adjust the number of old nonces that are cleaned up at shutdown.
+		 *
+		 * @since 9.0.0
+		 *
+		 * @param int $limit How many old nonces to remove at shutdown.
+		 */
+		$limit = apply_filters( 'jetpack_connection_nonce_cleanup_runtime_limit', static::CLEANUP_RUNTIME_LIMIT );
+
+		static::delete( $limit, time() - static::LIFETIME );
+
+		return true;
+	}
+
+
+	/**
+	 * Delete the nonces.
+	 *
+	 * @param int      $limit How many nonces to delete.
+	 * @param null|int $cutoff_timestamp All nonces added before this timestamp will be removed.
+	 *
+	 * @return int|false Number of removed nonces, or `false` if nothing to remove (or in case of a database error).
+	 */
+	public static function delete( $limit = 10, $cutoff_timestamp = null ) {
+		global $wpdb;
+
+		$ids = $wpdb->get_col(
+			$wpdb->prepare(
+				"SELECT option_id FROM `{$wpdb->options}`"
+				. " WHERE `option_name` >= 'jetpack_nonce_' AND `option_name` < %s"
```

Comment:
There's a logic error here, but it's unlikely to affect anything in practice for about 266 years.

If someone tries to pass a `$cutoff_timestamp` before 2001-09-09T01:46:40Z, it'll also delete most timestamps after that date. Say they want to delete all nonces before the year 2000, so they pass 946684800. But since as a string comparison "jetpack_nonce_1000000000" < "jetpack_nonce_946684800"...

The same sort of thing could happen in 2286 when timestamps reach 11 decimal digits.

But since WordPress didn't exist until 2003 (so there's no sane reason to set an earlier cutoff timestamp) and we'll hopefully have fixed nonces to use transients or something better than `wp_options` by 2286, we can probably live with this potential bug for now.

Reply:
> we'll hopefully have fixed nonces to use transients or something better than wp_options by 2286

The goal worthy of a noble man 😀